geoengine/permissions.py

```diff
@@ -0,0 +1,205 @@
+
+    def __str__(self) -> str:
+        return str(self.__role_id)
+
+    def __repr__(self) -> str:
+        return repr(self.__role_id)
+
+
+class UserId:
+    '''A wrapper for a role id'''
+
+    def __init__(self, user_id: UUID) -> None:
+        self.__user_id = user_id
+
+    @classmethod
+    def from_response(cls, response: Dict[str, str]) -> RoleId:
+        '''Parse a http response to an `UserId`'''
+        print(response)
+        if 'id' not in response:
+            raise GeoEngineException(response)
+
+        user_id = response['id']
+
+        return RoleId(UUID(user_id))
+
+    def __eq__(self, other) -> bool:
+        '''Checks if two role ids are equal'''
+        if not isinstance(other, self.__class__):
+            return False
+
+        return self.__user_id == other.__user_id  # pylint: disable=protected-access
+
+    def __str__(self) -> str:
+        return str(self.__user_id)
+
+    def __repr__(self) -> str:
+        return repr(self.__user_id)
+
+
+ADMIN_ROLE_ID: RoleId = RoleId(UUID("d5328854-6190-4af9-ad69-4e74b0961ac9"))
+REGISTERED_USER_ROLE_ID: RoleId = RoleId(UUID("4e8081b6-8aa6-4275-af0c-2fa2da557d28"))
+ANONYMOUS_USER_ROLE_ID: RoleId = RoleId(UUID("fd8e87bf-515c-4f36-8da6-1a53702ff102"))
+
+
+def add_permission(role: RoleId, resource: ResourceId, permission: Permission, timeout: int = 60):
+    """Add a permission to a resource for a role. Requires admin role."""
+
+    session = get_session()
+
+    payload = json.dumps({
+        "roleId": role,
+        "resourceId": resource,
+        "permission": permission
+    }, default=str)
+
+    headers = session.auth_header
+    headers['Content-Type'] = 'application/json'
+
+    response = req.put(
+        url=f'{session.server_url}/permissions',
+        headers=headers,
+        timeout=timeout,
+        data=payload
+    )
+
+    if not response.ok:
+        raise GeoEngineException(response.json())
+
+
+def remove_permission(role: UUID, resource: ResourceId, permission: Permission, timeout: int = 60):
+    """Removes a permission to a resource from a role. Requires admin role."""
+
+    session = get_session()
+
+    payload = json.dumps({
+        "roleId": role,
+        "resourceId": resource,
+        "permission": permission
+    }, default=str)
+
+    headers = session.auth_header
+    headers['Content-Type'] = 'application/json'
+
+    response = req.delete(
+        url=f'{session.server_url}/permissions',
+        headers=headers,
+        timeout=timeout,
+        data=payload
+    )
+
+    if not response.ok:
+        raise GeoEngineException(response.json())
+
+
+def add_role(name: str, timeout: int = 60) -> RoleId:
+    """Add a new role. Requires admin role."""
+
+    session = get_session()
+
+    headers = session.auth_header
+
+    response = req.put(
+        url=f'{session.server_url}/roles',
+        headers=headers,
+        timeout=timeout,
+        json={
+            "name": name
+        }
```

Comment:
there should be a type for that in api.py

Reply:
cb77f6e92122f0cc346dbc4ab2320a44a0ceea55